bucket.go

```diff
@@ -70,3 +80,484 @@ func (b *Bucket) Remove(ctx context.Context) error {
+// Parameters:
+//   - ctx: Context for cancellation and timeout control.
+//   - entry: Name of the entry to read from.
+//   - ts: Optional A UNIX timestamp in microseconds. If it is empty, the latest record is returned.
+//   - id: Optional A query ID to read the next record in the query. If it is set, the parameter ts is ignored.
+//   - head: If true, performs a HEAD request to fetch metadata only.
+//
+// It returns a readableRecord or an error if the read fails.
+//
+// Use readableRecord.Read() to read the content of the reader.
+func (b *Bucket) BeginRead(ctx context.Context, entry string, id *string, head bool) (*ReadableRecord, error) {
+	return b.readRecord(ctx, entry, id, head)
+}
+
+// readRecord prepares an entry record reader from the reductstore server.
+func (b *Bucket) readRecord(ctx context.Context, entry string, ts *string, head bool) (*ReadableRecord, error) {
+	query := url.Values{}
+	if ts != nil {
+		query.Set("ts", *ts)
+	}
+
+	path := fmt.Sprintf("/b/%s/%s?%s", b.Name, entry, query.Encode())
+
+	var req *http.Request
+	var err error
+	if head {
+		req, err = b.HTTPClient.NewRequestWithContext(ctx, http.MethodHead, path, nil)
+	} else {
+		req, err = b.HTTPClient.NewRequestWithContext(ctx, http.MethodGet, path, nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if !head {
+		req.Header.Set("Accept", "application/octet-stream")
+	}
+
+	resp, err := b.HTTPClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	errorMessage := resp.Header.Get("x-reduct-error")
+	if resp.StatusCode == http.StatusNoContent {
+		if errorMessage == "" {
+			errorMessage = "No content"
+		}
+		return nil, model.APIError{Status: http.StatusNoContent, Message: errorMessage}
+	}
+	// check there is data in the response
+	if resp.ContentLength == 0 || resp.Body == nil {
+		return nil, model.APIError{Status: http.StatusNoContent, Message: "No content"}
+	}
+
+	timeStr := resp.Header.Get("x-reduct-time")
+	sizeStr := resp.Header.Get("content-length")
+	last := resp.Header.Get("x-reduct-last") == "1"
+
+	labels := make(map[string]any)
+	for key, values := range resp.Header {
+		if strings.HasPrefix(key, "x-reduct-label-") {
+			labels[strings.TrimPrefix(key, "x-reduct-label-")] = values[0]
+		}
+	}
+
+	timeVal, _ := strconv.ParseInt(timeStr, 10, 64) //nolint:errcheck //not needed
+	sizeVal, _ := strconv.ParseInt(sizeStr, 10, 64) //nolint:errcheck //not needed
+	record := NewReadableRecord(timeVal, sizeVal, last, resp.Body, labels, resp.Header.Get("Content-Type"))
+	return record, nil
+
+}
+
+// BeginWrite starts a record writer for an entry.
+//
+// Parameters:
+//   - entry the name of the entry to write the record to.
+//   - options:
+//   - TimeStamp: timestamp in microseconds, it is set to current time if not provided
+//   - ContentType: "text/plain"
+//   - Labels: record label kev:value pairs  {label1: "value1", label2: "value2"}.
+func (b *Bucket) BeginWrite(entry string, options *WriteOptions) *WritableRecord {
```

Comment:
all other functions have the context in parameters. I think it makes sense to add it here even if we don't use it for now.